apps/mymedicare_cb/models.py

```diff
@@ -1,4 +1,87 @@
+import requests
+import logging
 from django.db import models
+from django.contrib.auth.models import User, Group
+from apps.accounts.models import UserProfile
+from apps.fhir.authentication import convert_sls_uuid
+from apps.fhir.bluebutton.models import Crosswalk
+from apps.fhir.bluebutton.utils import get_resourcerouter, FhirServerAuth
+
+__author__ = "Alan Viars"
+
+logger = logging.getLogger('hhs_server.%s' % __name__)
+
+
+def get_and_update_user(user_info):
+    try:
+        user = User.objects.get(username=convert_sls_uuid(user_info['sub']))
+        if not user.first_name:
+            user.first_name = user_info['given_name']
+        if not user.last_name:
+            user.last_name = user_info['family_name']
+        if not user.email:
+            user.email = user_info['email']
+        user.save()
+    except User.DoesNotExist:
+        # Create a new user. Note that we can set password
+        # to anything, because it won't be checked.
+        user = User(username=user_info['sub'][9:36], password='',
+                    first_name=user_info['given_name'],
+                    last_name=user_info['family_name'],
+                    email=user_info['email'])
+        user.save()
+    up, created = UserProfile.objects.get_or_create(
+        user=user, user_type='BEN')
+    group = Group.objects.get(name='BlueButton')
+    user.groups.add(group)
+    # Log in the user
+    user.backend = 'django.contrib.auth.backends.ModelBackend'
+
+    # Determine patient_id
+    fhir_source = get_resourcerouter()
+    crosswalk, g_o_c = Crosswalk.objects.get_or_create(
+        user=user, fhir_source=fhir_source)
+    hicn = user_info.get('hicn', "")
+    crosswalk.user_id_hash = hicn
+    crosswalk.save()
+
+    auth_state = FhirServerAuth(None)
+    certs = (auth_state['cert_file'], auth_state['key_file'])
+
+    # URL for patient ID.
+    url = fhir_source.fhir_url + \
+        "Patient/?identifier=http%3A%2F%2Fbluebutton.cms.hhs.gov%2Fidentifier%23hicnHash%7C" + \
```

Comment:
This is probably out of scope, but I think this can be switched to bluebutton.cms.gov, but you'd have to check with Karl

Reply:
@karlmdavis thoughts?